ui/core/sim_concurrent.ts

```diff
@@ -0,0 +1,167 @@
+		this.iterationsDone[idx] = msg.completedIterations;
+		this.dpsValues[idx] = msg.dps;
+		this.hpsValues[idx] = msg.hps;
+
+		if (msg.finalRaidResult) {
+			this.finalResults[idx] = msg.finalRaidResult;
+		}
+	}
+
+	makeProgressMetrics(): ProgressMetrics {
+		return ProgressMetrics.create({
+			totalIterations: this.iterationsTotal,
+			completedIterations: this.getIterationsDone(),
+			dps: this.getDpsAvg(),
+			hps: this.getHpsAvg(),
+		});
+	}
+}
+
+function makeAndSendErrorResult(err: string, onProgress: WorkerProgressCallback): RaidSimResult {
+	const errRes = RaidSimResult.create({ errorResult: err });
+	onProgress(ProgressMetrics.create({ finalRaidResult: errRes }));
+	console.error(err);
+	return errRes;
+}
+
+function splitRequest(request: RaidSimRequest, wp: WorkerPool): Promise<RaidSimRequestSplitResult> {
+	return wp.raidSimRequestSplit(RaidSimRequestSplitRequest.create({
+		splitCount: wp.getNumWorkers(),
+		request: request,
+	}));
+}
+
+function combineResults(results: RaidSimResult[], wp: WorkerPool): Promise<RaidSimResultCombinationResult> {
+	return wp.raidSimResultCombination(RaidSimResultCombinationRequest.create({
+		results: results,
+	}));
+}
+
+interface SimRunResult {
+	errorResult?: RaidSimResult;
+	results: RaidSimResult[];
+	progressMetricsFinal: ProgressMetrics;
+}
+
+async function runSims(requests: RaidSimRequest[], totalIterations: number, wp: WorkerPool, onProgress: WorkerProgressCallback): Promise<SimRunResult> {
+	let resolve: ((r: SimRunResult) => void) | undefined = undefined;
+
+	const csp = new ConcurrentSimProgress(requests.length, totalIterations);
+	let progressCounter = 0;
+	let running = requests.length;
+
+	const progressHandler = async (idx: number, pm: ProgressMetrics) => {
+		if (!resolve) return;
+
+		csp.updateProgress(idx, pm)
+
+		if (pm.finalRaidResult) {
+			let errRes: RaidSimResult | undefined;
+			if (pm.finalRaidResult.errorResult) {
+				console.error(`Worker ${idx} had an error!`);
+				errRes = pm.finalRaidResult;
+				// This sucks, but it's better than having long running workers forever.
+				if (requests[0].simOptions!.iterations > 1000) {
+					console.log("Terminating all workers to get going again.");
+					const num = wp.getNumWorkers();
+					wp.setNumWorkers(0);
+					wp.setNumWorkers(num);
+				}
+			}
+
+			running--;
+			if (errRes || running == 0) {
+				resolve({
+					errorResult: errRes,
+					results: csp.finalResults,
+					progressMetricsFinal: csp.makeProgressMetrics(),
+				});
+				resolve = undefined;
+			}
+		}
+
+		progressCounter++;
+		if (progressCounter % running == 0) {
+			onProgress(csp.makeProgressMetrics());
+		}
+	}
+
+	for (let i = 0; i < requests.length; i++) {
+		wp.raidSimAsync(requests[i], pm => { progressHandler(i, pm) });
+	}
+
+	return new Promise(res => {
+		resolve = res;
+	});
+}
+
+export async function runConcurrentSim(request: RaidSimRequest, workerPool: WorkerPool, onProgress: WorkerProgressCallback): Promise<RaidSimResult> {
```

Comment:
No need to explicetly define the return type, as it is inferred by default

Reply:
Still doesn't hurt to define them. Prevents wrong or missing returns and shows the error for the function, instead of down the line where the function is used.